src/Raven.Server/ServerWide/ClusterStateMachine.cs

```diff
@@ -1442,26 +1443,43 @@ public override void EnsureNodeRemovalOnDeletion(TransactionOperationContext con
             _parent.InsertToLeaderLog(context, term, context.ReadObject(djv, "remove"), RachisEntryFlags.StateMachineCommand);
         }
 
+        private readonly ConcurrentSet<long> _scheduledValueChangedNotificationIndexes = new ConcurrentSet<long>();
+
         private void NotifyValueChanged(TransactionOperationContext context, string type, long index)
         {
-            context.Transaction.InnerTransaction.LowLevelTransaction.OnDispose += transaction =>
+            //in case of batches that do multiple changes with the same index, we need to notify about those changes only once
+            if (_scheduledValueChangedNotificationIndexes.TryAdd(index))
```

Comment:
What if you have a batch command that contains 3 commands.
After committing every command does something asyncly (`ExecuteAsyncTask`).
So you will notify the waiter that you are done only after the first command is completed. not all three, right?

Reply:
All of them should happen in the same tx, no?
In that case, we'll register for this only once, and notify once the tx is completely committed

Reply:
After the tx is committed, we run an async task.
So I have a command to notify db1 and another command (with the same index) to notify db2.
In this case, db2 wouldn't be notified.

Can't we use the `ExecuteManyOnDispose` for this purpose?

Reply:
_can_ we have a batch command on mulitiple databases?

Reply:
I don't know if we are using it this way. 
But I don't see a reason why not.